src/client/mod.rs

```diff
@@ -1,260 +1,190 @@
-            }
-        };
+// ===== impl Builder =====
 
-        loop {
-            let message = try_ready!(body.poll()
-                .map_err(|e| match e {
-                    ::Error::Inner(h2) => ::Error::Grpc(::Status::from(h2)),
-                    ::Error::Grpc(err) => ::Error::Grpc(err),
-                }));
-
-            match (self.message.is_some(), message) {
-                (false, Some(msg)) => {
-                    self.message = Some(msg);
-                    continue;
-                },
-                (true, None) => {
-                    let head = self.head.take().expect("polled more than once");
-                    let body = self.message.take().expect("polled more than once");
-                    let http = http::Response::from_parts(head, body);
-                    let resp = ::Response::from_http(http);
-                    return Ok(Async::Ready(resp));
-                }
-                (true, Some(_)) => {
-                    debug!("Unary decoder found 2 messages");
-                    return Err(::Error::Grpc(Status::UNKNOWN));
-                }
-                (false, None) => {
-                    debug!("Unary decoder ended before any messages");
-                    return Err(::Error::Grpc(Status::UNKNOWN));
-                }
-            }
-        }
+impl Builder {
+    /// Return a new client builder
+    pub fn new() -> Self {
+        Self::default()
     }
-}
 
-// ====== impl Stream =====
+    pub fn uri(&mut self, value: uri::Uri) -> &mut Self {
+        // TODO: Validate that the path is empty. Doing this is waiting for
+        // hyper/http#149.
 
-impl<F, B> Streaming<F, B>
-where
-    F: Future<Item=::Response<B>>,
-    B: Stream<Error=::Error<::h2::Error>>,
-{
-    pub fn map_future(future: F) -> Self {
-        Streaming {
-            body: None,
-            future,
-        }
+        self.uri = Some(value);
+        self
     }
-}
 
-impl<F, B, E> Stream for Streaming<F, B>
-where
-    F: Future<Item=::Response<B>, Error=::Error<E>>,
-    B: Stream<Error=::Error<::h2::Error>>,
-{
-    type Item = B::Item;
-    type Error = ::Error<E>;
-
-    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
-        loop {
-            if let Some(ref mut body) = self.body {
-                return body.poll().map_err(|e| match e {
-                        ::Error::Inner(h2) => ::Error::Grpc(::Status::from(h2)),
-                        ::Error::Grpc(err) => ::Error::Grpc(err),
-                    });
-            } else {
-                let res = try_ready!(self.future.poll());
-                self.body = Some(res.into_http().into_parts().1);
-            }
-        }
-    }
-}
+    pub fn build<T>(&mut self, inner: T) -> Grpc<T>
+    where T: HttpService,
+    {
+        let uri = self.uri.as_mut().expect("service URI is required");
+
+        let scheme = uri.scheme_part()
+            .expect("gRPC service URI must include scheme and authority")
```

Comment:
Should we be panicking or returning a `Result`?

Most cases, a user probably constructs a Grpc by hand and so a bad URI is a programmer error, though it could be that it is constructed by config or user input...

Reply:
A result here is probably fine.